### PR TITLE
CDAP-2464 Documentation to highlight different CDAP modes

### PR DESCRIPTION
**Pull request by awholegunch**
Add text and links about different CDAP modes available.

Passes [Quick Build](http://builds.cask.co/browse/CDAP-DOB52-3). (Updated for Build 3.)

Pages Of Interest:
- [Overview](http://builds.cask.co/artifact/CDAP-DOB52/shared/build-3/Docs-HTML/3.3.0-SNAPSHOT/en/index.html)
- [Installation](http://builds.cask.co/artifact/CDAP-DOB52/shared/build-3/Docs-HTML/3.3.0-SNAPSHOT/en/admin-manual/installation/index.html)
- [Testing a CDAP Application](http://builds.cask.co/artifact/CDAP-DOB52/shared/build-3/Docs-HTML/3.3.0-SNAPSHOT/en/developers-manual/testing/testing.html#test-cdap)

Fix for https://issues.cask.co/browse/CDAP-2464

### CODE REVIEW

File: cdap-docs/_common/_source/index.rst [hunk 1]
```diff
@@ -37,6 +37,11 @@ and accessing core CDAP services. CDAP defines and implements a diverse collecti
 services that support applications and data on existing Hadoop infrastructure such as
 HBase, HDFS, YARN, MapReduce, Hive, and Spark.
 
+CDAP can be run in different modes: in-memory mode for unit testing, standalone CDAP for
+testing on a developer's laptop, and distributed CDAP for staging and production.
+Regardless of the runtime edition, CDAP is fully functional and the code you develop never
```
Comment (bdmogal) [inline cdap-docs/_common/_source/index.rst]: runtime mode?

File: cdap-docs/_common/_source/index.rst [hunk 2]
```diff
@@ -37,6 +37,11 @@ and accessing core CDAP services. CDAP defines and implements a diverse collecti
 services that support applications and data on existing Hadoop infrastructure such as
 HBase, HDFS, YARN, MapReduce, Hive, and Spark.
 
+CDAP can be run in different modes: in-memory mode for unit testing, standalone CDAP for
```
Comment (bdmogal) [inline cdap-docs/_common/_source/index.rst]: unit testing/CI

**Comment by bdmogal** [comment]
@awholegunch do you think we need more information for this section? Here is something I had used recently: 

## Modes of operation
### In-Memory:
* Purpose built for writing unit tests, CI pipelines
* Mimicks storage technologies as in-memory data structures, e.g. Java NavigableMap 
* Uses Java Threads as the processing abstraction (via Apache Twill)
### Standalone/SDK:
* Designed to run in a standalone environment, for development and test
* Uses LevelDB/Local File System as the storage technology
* Uses Java Threads as the processing abstraction (via Apache Twill)
### Distributed:
* Production/Staging/QA mode, runs in a distributed environment
* Uses Apache HBase and HDFS as the storage technology today
* Uses Apache Yarn Containers as the processing abstraction (via Apache Twill)


CDAP Applications are developed against the CDAP APIs, making the switch between these modes seamless. An application developed using a given mode can easily run in another mode.

File: cdap-docs/_common/_source/index.rst (hunk 1 above)
Comment (awholegunch) [inline cdap-docs/_common/_source/index.rst]: Fixed (in two files.)

File: cdap-docs/_common/_source/index.rst (hunk 2 above)
Comment (awholegunch) [inline cdap-docs/_common/_source/index.rst]: Fixed (in two files.)

**Comment by awholegunch** [comment]
"Mimicks storage technologies" -> "Mimics storage technologies"

**Comment by awholegunch** [comment]
I've added material from that set of three points as appropriate.

File: cdap-docs/_common/_source/index.rst [hunk 3]
```diff
@@ -93,7 +98,8 @@ on the installation, monitoring and diagnosing fully distributed CDAP in a Hadoo
     guarantees, transaction service maintenance, troubleshooting and introduces the CDAP UI
   - **Troubleshooting and Appendices:** Installation and configuration troubleshooting
     tips, and appendices that cover the CDAP installation and security configuration files
-
+  - **Modes:** CDAP can be run in-memory for unit testing, standalone for testing on a developer's
```
Comment (bdmogal) [inline cdap-docs/_common/_source/index.rst]: standalone for development and testing...?

Comment (bdmogal) [inline cdap-docs/_common/_source/index.rst]: Although, this seems to be a repeat of lines 40-42?

File: cdap-docs/admin-manual/source/installation/index.rst [hunk 4]
```diff
@@ -50,25 +50,37 @@ Installation
...
-underlying Big Data infrastructure is emulated using in-memory data structures and there
-is no persistence. The CDAP UI is not available in this mode.
+The in-memory CDAP allows you to easily run CDAP for use in unit tests and continuous
+integration (CI) pipelines. In this mode, the underlying Big Data infrastructure is
+emulated using in-memory data structures and there is no persistence. The CDAP UI is not
+available in this mode. See :ref:`test-cdap` for information and examples on using this
+mode.
+
+- Purpose-built for writing unit tests and CI pipelines
+- Mimics storage technologies as in-memory data structures; for example, Java NavigableMap
```
Comment (bdmogal) [inline cdap-docs/admin-manual/source/installation/index.rst]: perhaps a link for NavigableMap?

File: cdap-docs/admin-manual/source/installation/index.rst [hunk 5]
```diff
@@ -50,25 +50,37 @@ Installation
...
-The in-memory CDAP allows you to easily run CDAP for use in unit tests. In this mode, the
-underlying Big Data infrastructure is emulated using in-memory data structures and there
-is no persistence. The CDAP UI is not available in this mode.
+The in-memory CDAP allows you to easily run CDAP for use in unit tests and continuous
+integration (CI) pipelines. In this mode, the underlying Big Data infrastructure is
+emulated using in-memory data structures and there is no persistence. The CDAP UI is not
+available in this mode. See :ref:`test-cdap` for information and examples on using this
+mode.
+
+- Purpose-built for writing unit tests and CI pipelines
```
Comment (bdmogal) [inline cdap-docs/admin-manual/source/installation/index.rst]: these points seem sort of abrupt. Do you think they need a section header? Features?

File: cdap-docs/_common/_source/index.rst (hunk 3 above)
Comment (awholegunch) [inline cdap-docs/_common/_source/index.rst]: Good idea. I have edited down line ~101, and moved the "standalone for development and testing" up to line 40-42. Please re-review...

File: cdap-docs/admin-manual/source/installation/index.rst (hunk 4 above)
Comment (awholegunch) [inline cdap-docs/admin-manual/source/installation/index.rst]: Added.

File: cdap-docs/admin-manual/source/installation/index.rst (hunk 5 above)
Comment (awholegunch) [inline cdap-docs/admin-manual/source/installation/index.rst]: Good point. I will re-work this area.

**Comment by awholegunch** [comment]
Closing and reopening.

**Pull request closed without merging**